pallets/amm/src/lib.rs

```diff
@@ -273,77 +289,73 @@ pub mod pallet {
-            let pool_liquidity_amount: PoolLiquidityAmount = Self::pools(base_asset, quote_asset);
-
-            ensure!(
-                pool_liquidity_amount.ownership >= ownership_to_remove,
-                Error::<T, I>::MoreLiquidity
-            );
-
-            let base_amount = (ownership_to_remove
-                .saturating_mul(pool_liquidity_amount.base_amount))
-            .checked_div(pool_liquidity_amount.ownership)
-            .expect("cannot overflow with positive divisor; qed");
-
-            let quote_amount = (ownership_to_remove
-                .saturating_mul(pool_liquidity_amount.quote_amount))
-            .checked_div(pool_liquidity_amount.ownership)
-            .expect("cannot overflow with positive divisor; qed");
-
             Pools::<T, I>::try_mutate(
                 &base_asset,
                 &quote_asset,
                 |pool_liquidity_amount| -> DispatchResult {
-                    pool_liquidity_amount.base_amount = pool_liquidity_amount
+                    let mut liquidity_amount = pool_liquidity_amount
+                        .take()
+                        .ok_or(Error::<T, I>::PoolDoesNotExist)?;
+                    ensure!(
+                        liquidity_amount.ownership >= ownership_to_remove,
+                        Error::<T, I>::MoreLiquidity
+                    );
+
+                    let base_amount = (ownership_to_remove
+                        .saturating_mul(liquidity_amount.base_amount))
+                    .checked_div(liquidity_amount.ownership)
+                    .expect("cannot overflow with positive divisor; qed");
+
+                    let quote_amount = (ownership_to_remove
+                        .saturating_mul(liquidity_amount.quote_amount))
+                    .checked_div(liquidity_amount.ownership)
+                    .expect("cannot overflow with positive divisor; qed");
+
+                    liquidity_amount.base_amount = liquidity_amount
                         .base_amount
                         .checked_sub(base_amount)
                         .ok_or(ArithmeticError::Underflow)?;
-                    pool_liquidity_amount.quote_amount = pool_liquidity_amount
+                    liquidity_amount.quote_amount = liquidity_amount
                         .quote_amount
                         .checked_sub(quote_amount)
                         .ok_or(ArithmeticError::Underflow)?;
-                    pool_liquidity_amount.ownership = pool_liquidity_amount
+                    liquidity_amount.ownership = liquidity_amount
                         .ownership
                         .checked_sub(ownership_to_remove)
                         .ok_or(ArithmeticError::Underflow)?;
-                    Ok(())
-                },
-            )?;
+                    *pool_liquidity_amount = Some(liquidity_amount.clone());
+
+                    LiquidityProviders::<T, I>::try_mutate(
+                        (who.clone(), base_asset, quote_asset),
```

Comment:
can this clone be removed?

Reply:
Its needed here. It can't be removed.

Reply:
wont `&who` be enough?

Reply:
Yes @GopherJ this would work. I have modified. Please review